# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wgpu/wgpu-core/src/command/draw.rs
# Commit: ef5dc3e04e5f
# Full Hash: ef5dc3e04e5f271eea0636ab3a495e95cc912f1d
# Author: Dzmitry Malyshau <dmalyshau@mozilla.com>
# Date: 2021-09-04 09:40:24
# Regressor Bug: 1726626
# File Overlap Count: 3
# Description:
#   Bug 1726626 - Move gfx/wgpu into a 3rd party dependency r=jgilbert,bholley
#   
#   This update makes wgpu a vendored dependency instead of having it in gfx/wgpu.
#   
#   ## Notes
# ==============================================================================

diff --git a/gfx/wgpu/wgpu-core/src/command/draw.rs b/third_party/rust/wgpu-core/src/command/draw.rs
rename from gfx/wgpu/wgpu-core/src/command/draw.rs
rename to third_party/rust/wgpu-core/src/command/draw.rs
--- a/gfx/wgpu/wgpu-core/src/command/draw.rs
+++ b/third_party/rust/wgpu-core/src/command/draw.rs
@@ -1,28 +1,24 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-
 /*! Draw structures - shared between render passes and bundles.
 !*/
 
 use crate::{
     binding_model::PushConstantUploadError,
+    error::ErrorFormatter,
     id,
-    resource::BufferUse,
     track::UseExtendError,
     validation::{MissingBufferUsageError, MissingTextureUsageError},
 };
 use wgt::{BufferAddress, BufferSize, Color};
 
 use std::num::NonZeroU32;
 use thiserror::Error;
 
-pub type BufferError = UseExtendError<BufferUse>;
+pub type BufferError = UseExtendError<hal::BufferUses>;
 
 /// Error validating a draw call.
 #[derive(Clone, Debug, Error, PartialEq)]
 pub enum DrawError {
     #[error("blend constant needs to be set")]
     MissingBlendConstant,
     #[error("render pipeline must be set")]
     MissingPipeline,
@@ -72,20 +68,20 @@ pub enum RenderCommandError {
     #[error("dynamic buffer offset {0} does not respect `BIND_BUFFER_ALIGNMENT`")]
     UnalignedBufferOffset(u64),
     #[error("number of buffer offsets ({actual}) does not match the number of dynamic bindings ({expected})")]
     InvalidDynamicOffsetCount { actual: usize, expected: usize },
     #[error("render pipeline {0:?} is invalid")]
     InvalidPipeline(id::RenderPipelineId),
     #[error("QuerySet {0:?} is invalid")]
     InvalidQuerySet(id::QuerySetId),
-    #[error("Render pipeline is incompatible with render pass")]
-    IncompatiblePipeline(#[from] crate::device::RenderPassCompatibilityError),
-    #[error("pipeline is not compatible with the depth-stencil read-only render pass")]
-    IncompatibleReadOnlyDepthStencil,
+    #[error("Render pipeline targets are incompatible with render pass")]
+    IncompatiblePipelineTargets(#[from] crate::device::RenderPassCompatibilityError),
+    #[error("pipeline writes to depth/stencil, while the pass has read-only depth/stencil")]
+    IncompatiblePipelineRods,
     #[error("buffer {0:?} is in error {1:?}")]
     Buffer(id::BufferId, BufferError),
     #[error("buffer {0:?} is destroyed")]
     DestroyedBuffer(id::BufferId),
     #[error(transparent)]
     MissingBufferUsage(#[from] MissingBufferUsageError),
     #[error(transparent)]
     MissingTextureUsage(#[from] MissingTextureUsageError),
@@ -93,16 +89,33 @@ pub enum RenderCommandError {
     PushConstants(#[from] PushConstantUploadError),
     #[error("Invalid Viewport parameters")]
     InvalidViewport,
     #[error("Invalid ScissorRect parameters")]
     InvalidScissorRect,
     #[error("Support for {0} is not implemented yet")]
     Unimplemented(&'static str),
 }
+impl crate::error::PrettyError for RenderCommandError {
+    fn fmt_pretty(&self, fmt: &mut ErrorFormatter) {
+        fmt.error(self);
+        match *self {
+            Self::InvalidBindGroup(id) => {
+                fmt.bind_group_label(&id);
+            }
+            Self::InvalidPipeline(id) => {
+                fmt.render_pipeline_label(&id);
+            }
+            Self::Buffer(id, ..) | Self::DestroyedBuffer(id) => {
+                fmt.buffer_label(&id);
+            }
+            _ => {}
+        };
+    }
+}
 
 #[derive(Clone, Copy, Debug, Default)]
 #[cfg_attr(
     any(feature = "serial-pass", feature = "trace"),
     derive(serde::Serialize)
 )]
 #[cfg_attr(
     any(feature = "serial-pass", feature = "replay"),
@@ -149,17 +162,17 @@ pub enum RenderCommand {
     SetViewport {
         rect: Rect<f32>,
         //TODO: use half-float to reduce the size?
         depth_min: f32,
         depth_max: f32,
     },
     SetScissor(Rect<u32>),
     SetPushConstant {
-        stages: wgt::ShaderStage,
+        stages: wgt::ShaderStages,
         offset: u32,
         size_bytes: u32,
         /// None means there is no data and the data should be an array of zeros.
         ///
         /// Facilitates clears in renderbundles which explicitly do their clears.
         values_offset: Option<u32>,
     },
     Draw {